### DISCUSSION TITLE
Question about the FullscreenWindowMode

### DISCUSSION THREAD

**Discussion by HerMajestyDrMona** [elishacloud/dxwrapper#161] [Q&A]
Hi.

I have a question about the `FullscreenWindowMode` in DX9 (using D8to9).

I noticed that sometimes in this mode, the Windows taskbar doesn't hide. I also noticed, that the game window will not minimize on ALT+TAB switching. I know that it's rather an expected behavior.

Some games that support the "Borderless fullscreen windowed mode" in their settings (e.g. awful example: League of Legends) can respond to ALT+TAB and don't have the visible taskbar problem. What is the difference? Were these games programmed to minimize the window on ALT+TAB, or they use some different mode than the one available in dxwrapper?

**Comment by mirh** [comment]
Pretty sure no game specifically addresses ALT+TAB. 
You are probably experiencing some of the dozens possible interactions between windows compatibility shims, and fullscreen optimizations. 
EDIT: ddraw [may also](https://www.vogons.org/viewtopic.php?f=8&t=47772&start=40) play a role

**Comment by HerMajestyDrMona** [comment]
Thanks for the reply.

Seems like it was the other way around. The game I support (EE2) has some functions that are pausing the game window when it becomes inactive. There are also separate functions that are muting the audio when the window loses focus (even in the windowed mode). I fixed these issues with a new option that allows my injected DLL to modify game memory and fix these things when the game is in the windowed mode (and revers when it switches back to fullscreen on ALT+ENTER).

For now I also resigned from replacing d3d8to9 to dxwrapper, because I fixed the borderless windowed mode by the assembly myself. The game allows the windowed mode and switching between fullscreen and windowed mode with a hotkey. For the windowed mode I hijack the style and modify it to remove ` ~(WS_CAPTION | WS_THICKFRAME | WS_MINIMIZEBOX | WS_MAXIMIZEBOX | WS_SYSMENU)`. I was also able to support multiple displays by modifying the X, Y coordinates, depending on selected adapter window. In the past, I also made a hotkey that allows the mouse cursor to leave the game window (constrain cursor). So now, I'm finally satisfied how this game works on multiple monitors!

It's still tempting to migrate to dxwrapper for `AnisotropicFiltering`, `ForceMixedVertexProcessing` and other useful options, but I'm not sure if it's better for the performance and stability.

**Reply by mirh** [reply]
Oh, so that's why I seemed to remember you. 
Alas the options from #111 still aren't doing a thing for me (putting aside I don't really think you'd need all that much of a bump for a 2005 RTS). 
And TIL dxwrapper doesn't inherently use d3d8to9.

**Reply by HerMajestyDrMona** [reply]
Yeah, I detected that enabling `ForceSystemMemVertexCache` actually reduces performance a lot (huge FPS loss). Not sure about the `ForceMixedVertexProcessing`, since it doesn't make any difference to me. Maybe I'll better not touch options that are not needed.

I read the issue you linked, there was a discussion about the future `ForceHardwareVertexProcessing` option. Do you think the game by any chance uses the hardware vertex processing already?

**Reply by HerMajestyDrMona** [reply]
Nevermind, it seems like it's already using HARDWARE. I simply added a message box under `else if (BehaviorFlags & D3DCREATE_HARDWARE_VERTEXPROCESSING)` and it shows up :D

**Comment by elishacloud** [comment]
dxwrapper does not do anything special with the `FullscreenWindowMode` in DX9.  Basically it just changes the following:

1. It changes the call into `CreateDevice` to tell DirectX to use [windowed mode](https://github.com/elishacloud/dxwrapper/blob/392335a574d8a8c077e2f2a0eb9b59fe8490945e/d3d9/IDirect3D9Ex.cpp#L495-L499) rather than fullscreen mode.
2. It updates the [window style and location](https://github.com/elishacloud/dxwrapper/blob/392335a574d8a8c077e2f2a0eb9b59fe8490945e/d3d9/IDirect3D9Ex.cpp#L563-L622) so that the window will appear in the correct size and location on the screen.
3. It changes the [resolution of the system](https://github.com/elishacloud/dxwrapper/blob/392335a574d8a8c077e2f2a0eb9b59fe8490945e/d3d9/IDirect3D9Ex.cpp#L514-L536) to match the resolution of the game.

> Some games that support the "Borderless fullscreen windowed mode" in their settings (e.g. awful example: League of Legends) can respond to ALT+TAB and don't have the visible taskbar problem. What is the difference? Were these games programmed to minimize the window on ALT+TAB, or they use some different mode than the one available in dxwrapper?

dxwrapper does not do anything to handle this.  I guess this may be the way the game works.  More code would need to be added to solve this issue.

**Reply by HerMajestyDrMona** [reply]
Thanks for the reply! I actually had to include dozens of assembly fixes to the game to make it render correctly in the borderless windowed mode and secondary monitors. Even `EnumDisplaySettings` was called by the game with the NULL parameter, so it was only allowing the primary monitor screen resolutions on secondary monitors... Also taking a screenshot (game feature) was resulting in crashes or incorrect output. I actually "hacked" d3d8to9 to intercept calls of the screenshoting function (`CreateImageSurface`), save the screenshot to the correct name in JPG, PNG or BMP quality, and then return `D3DERR_INVALIDCALL` in the next call to `GetFrontBuffer`. This way the game doesn't take a screenshot using its old functions, and I can save it to the right folder, with the right (respected by user settings) quality, from any display adapter :) 
There was also an issue that the game was interrupting other game's "constrain cursor" lock. So to fix that, in my watchdog thread I detect if the game is iconic or the window inactive. then compare HMONITOR of the game window HWND with the monitor from the mouse position, and then `WriteProcessMemory` with _asm fixes that temporarily disable  interception of the mouse pointer. It gets reverted when the mouse cursor returns to the game window's monitor, or the game window gets focus again.
There were few more things like this, but the final result is awesome.

I'd like to ask you, have you by any chance heard about `D3D9On12`? It seems like the newest Intel GPUs force it for all DX9 games. It causes a bug in EE2, where the text of the menu buttons is not displaying... [Example](https://user-images.githubusercontent.com/24290461/197898630-12a19b70-52e0-4841-9d1c-4996cb42f81f.jpg). Since I don't have a new Intel laptop, do you think it's possible to somehow force d3d8to9 to use D3D9On12 to test this issue? I know that `D3D9On12` is for graphics drivers, but just wanted to ask if you are aware of any possible workaround.

**Reply by elishacloud** [reply]
> I'd like to ask you, have you by any chance heard about `D3D9On12`?

Yes.  You can simply call a different export in the d3d9.dll file.  See [here](https://devblogs.microsoft.com/directx/coming-to-directx-12-d3d9on12-and-d3d11on12-resource-interop-apis/).

**Reply by HerMajestyDrMona** [reply]
WOW! So I only need to add the first `typedef HRESULT` and `typedef IDirect3D9*`? or I need to rewrite all d3d8to9 functions to call D3D9On12?

**Reply by elishacloud** [reply]
Yeah, you just need to add the typedef and then call the new export.  The rest of the d3d8to9 code can remain the same.

Here are two sample implementations:
* See [here](https://github.com/FunkyFr3sh/cnc-ddraw/commit/5616af006b14df2b2e3a683c988de7f66371f188)
* https://github.com/narzoul/ForceD3D9On12

**Reply by HerMajestyDrMona** [reply]
When I added `typedef struct _D3D9ON12_ARGS` I got the `C2011	'_D3D9ON12_ARGS': 'struct' type redefinition` error during compilation. It compiles without it. Maybe it's already in `D3D9on12.h`?

**Reply by elishacloud** [reply]
Yeah, you don't need to add the struct.

**Reply by HerMajestyDrMona** [reply]
Sorry for being a newbie (I'm self learner). So if I understood correctly, I only need to add the two `typedef HRESULT` and `typedef IDirect3D9*` and that's all? Nothing else will be needed? I suppose it will require two different DLLs for d3d9 with D3D9On12 enabled and without? Because it can't be dynamically changed with a BOOL?

**Reply by mirh** [reply]
I'd be surprised if the thing worked out of the box with any wrapper, given how [surreptitiously](https://www.vogons.org/viewtopic.php?f=8&t=86912#p1098648) it just takes care of the "rendering behind the scenes" AFAICT. 
And indeed: https://github.com/narzoul/DDrawCompat/commit/20912dce37b481cd7e76eec90a37235ca026ecfe https://github.com/crosire/reshade/commit/7dd91fe58858119abe378d3687618963aece57d3

**Reply by HerMajestyDrMona** [reply]
Ok, maybe I'm just tired, but for some reasons I'm still having problems understanding it. Will I actually need to rewrite D3D8to9 calls to D3D9On12? You wrote that `The rest of the d3d8to9 code can remain the same`, and that I only need to `call the new export`? The new export is: `PFN_Direct3DCreate9On12`? Do I call it in the `Direct3DCreate8` function? Like, replace: `IDirect3D9 *const d3d = Direct3DCreate9(D3D_SDK_VERSION);` to `IDirect3D9 *const d3d = Direct3DCreate9On12(D3D_SDK_VERSION);` ?

**Reply by HerMajestyDrMona** [reply]
Ahhh!!! I'm stupid. Why I even bothered with including it, if I could simply paste d3d9.dll from ForceD3D9On12 Releases to the game folder... This way d3d8to9 DLL will use D3D9On12. 

And indeed, there is a problem with fonts not displaying in game! So that's what is happening to new Intel laptop users... I wonder how am I supposed to fix it, if it's a problem from Microsoft part :/

**Reply by elishacloud** [reply]
I added all the code needed for the native d3d8to9 to use D3D9on12.  You can see the code [here](https://github.com/elishacloud/d3d8to9/commit/ea6c31849a41013c46f66aa4db3373bced6a6231).

Note: if you have added headers you may not need all the code I added.  This is just for the native (unchanged) d3d8to9.

**Reply by HerMajestyDrMona** [reply]
WOW, thank you! Seems to work flawlessly. I also created an issue on D3D9on12 GitHub: https://github.com/microsoft/D3D9On12/issues/57
Wonder if they'll ever focus on it...

Can you by any chance guess what DX function calls are responsible for text displaying? I would then maybe add some more debug output and see what's going on...

**Reply by elishacloud** [reply]
In DirectX9 displaying text could be done by this:  https://github.com/apitrace/dxsdk/blob/master/Include/d3dx9effect.h

Edit: updated the link to the right one...

**Reply by HerMajestyDrMona** [reply]
Thanks. So this could be `SetString`, but it can't be found in d3d8to9? So maybe it's `SetTexture`? Either way, I think I wouldn't be able to workaround it. It seems to just be a bug in D3D9On12, and I bet they'll never fix it :D

**Reply by elishacloud** [reply]
These are done inside the d3dx9.dll file.  You can check if this file is used.  All of these calls (including `SetString`) ultimately call d3d9 functions, I believe.  I just don't know which d3d9 functions they are translated into.

**Reply by HerMajestyDrMona** [reply]
@elishacloud If you get a chance, could you please take a look at apitraces which I posted here?: https://github.com/microsoft/D3D9On12/issues/57#issuecomment-1292746990 ? With your huge experience maybe you would notice something useful.

**Reply by HerMajestyDrMona** [reply]
Interesting. It seems like there might be something wrong in `Direct3DDevice8::SetVertexShader` -> `hr = ProxyInterface->SetFVF(Handle)` function. When I added `Handle = D3DFVF_XYZ | D3DFVF_DIFFUSE | D3DFVF_TEX1;` before it, then the fonts are not appearing even on the normal d3d8to9 version. I guess it's a good starting point.
